nx/lib/nx.ex

```diff
@@ -3551,6 +3551,165 @@ defmodule Nx do
     end)
   end
 
+  @doc """
+  Transforms the leading dimension in a tensor into a vectorized axis.
+
+  Nested vectorizations will "remove" the leading axis from the shape
+  and append it to the vectorized axes
+
+  {2, 3, 4}
+  |> vectorize(:first) # => {3, 4} vec [first: 2]
+  |> vectorize(:second) # => {4} vec [first: 2, second: 3]
+  """
+  @doc type: :shape
+  def vectorize(tensor, name)
+
+  def vectorize(_tensor, name) when not is_atom(name) do
+    raise ArgumentError, "name for new vectorized axis must be an atom, got: #{inspect(name)}"
+  end
+
+  def vectorize(%Nx.Tensor{shape: {}}, _name) do
+    raise ArgumentError, "cannot vectorize tensor of rank 0"
+  end
+
+  def vectorize(%Nx.Tensor{names: names, shape: shape, vectorized_axes: vec_axes} = tensor, name) do
+    size = elem(shape, 0)
+    new_shape = Tuple.delete_at(shape, 0)
+    names = Enum.drop(names, 1)
+
+    %Nx.Tensor{
+      tensor
+      | shape: new_shape,
+        names: names,
+        vectorized_axes: vec_axes ++ [{name, size}]
+    }
+  end
+
+  defp unvectorize(%T{shape: shape, names: names, vectorized_axes: vectorized_axes} = tensor) do
+    {vectorized_names, vectorized_sizes} = Enum.unzip(vectorized_axes)
+
+    output_shape_l = vectorized_sizes ++ Tuple.to_list(shape)
+    output_shape = List.to_tuple(output_shape_l)
+
+    output_names = List.duplicate(nil, length(vectorized_sizes)) ++ names
+
+    output = %{tensor | shape: output_shape, names: output_names, vectorized_axes: []}
+    {output, vectorized_names}
+  end
+
+  defp unvectorize(
+         %T{shape: left_shape, names: left_shape_names} = left,
+         %T{shape: right_shape, names: right_shape_names} = right
+       ) do
+    left_names = Keyword.keys(left.vectorized_axes) |> MapSet.new()
+    right_names = Keyword.keys(right.vectorized_axes) |> MapSet.new()
+
+    left_only = MapSet.difference(left_names, right_names)
+    right_only = MapSet.difference(right_names, left_names)
+    in_both = MapSet.intersection(left_names, right_names)
```

Comment:
You can probably do all of this in a single pass. Convert the left (or right) side to a map. Then do a reduce collecting what exists on both, removing from the map, and what exists on one side. What remains on the map is what exists only on the other side.